src/history/hs_rec.c

```diff
@@ -794,9 +810,22 @@ __wt_hs_delete_key(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_t btre
         WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_start_ts, &hs_counter));
         ++hs_counter;
     }
+    /* Get the history store cursor position */
+    WT_ERR_NOTFOUND_OK(__hs_cursor_pos(session, hs_cursor, btree_id, key, WT_TS_NONE,
+                         error_on_ts_ordering, NULL, &non_ts_updates),
+      true);
 
-    WT_ERR(__hs_delete_reinsert_from_pos(session, hs_cursor, btree_id, key, WT_TS_NONE, reinsert,
-      true, error_on_ts_ordering, &hs_counter, NULL));
+    /*
+     * If we have non-timestamped updates, skip the reinserts as we do not need to zero the
+     * timestamps already at zero. If we have non-timestamped deletes (then reinsert is false)
+     * continue into the function to delete only. If we do not have a valid cursor position then
+     * nothing is found, return.
+     */
+    if (!(reinsert && non_ts_updates) && ret == 0)
```

Comment:
I realised that it might be better to have an if statement that directly does a goto done for code readability purposes.

```
    if (reinsert && non_ts_updates)
        goto done;
        
    if (ret == 0)
        WT_ERR(__hs_delete_reinsert_from_pos(session, hs_cursor, btree_id, key, WT_TS_NONE,
          reinsert, true, non_ts_updates, error_on_ts_ordering, &hs_counter, NULL));
    else
        ret = 0;
```
What do you think? It clears the nasty if condition, and makes a lot more sense to me.

Reply:
Made a few changes but I like breaking apart the complicated if statement - adjustments made here [2b180e8](https://github.com/wiredtiger/wiredtiger/pull/10686/commits/2b180e83db0c5ee2f52a7de9ce66b5a63924bdf0)